### ISSUE TITLE
[Issue] Poor performance in low-end devices

### ISSUE THREAD

**Issue by Terkyz** [MM2-0/Kvaesitso#257]
Kvaesitso works great and with each update it gets faster, but at least on my device (Samsung A10, 2GB), in the app drawer the performance is noticeably worse than in the main menu of the launcher, and in fact gives the feeling that it moves slower or is "heavier" when moving the finger.

I have noticed that the more apps you have, the greater the feeling of "heaviness" and is that the horizontal line where applications appear is noted that takes longer to load, because you see how it appears and apps are accommodated, while with the apps above this does not happen.

I have made several comparisons (although this does not mean anything, because not all launchers are made equal) and I do not have this problem with all launchers, but with some, and the common factor of "heaviness" is always the number of installed apps.

[Here](https://imgur.com/a/NNGuBi6) is a small example of the "heaviness".

**Comment by htaige** [comment]
I am experiencing the same issue on my Galaxy s10+. It is especially bad after booting the device as the launcher loads all the icons for the apps. The scrolling performance for the app drawer and the transition swipe to access the app drawer feel noticeably more sluggish when compared to other launchers.

**Comment by jspivey40** [comment]
Basically same thing here. Stuttering and hiccups or heaviness in many areas, and it does seem like it's worse with more apps installed. This is one of my favorite launchers, but the heaviness is a bit much for my stylus 5g

**Comment by Terkyz** [comment]
Title corrected.

**Comment by montchr** [comment]
Same here on Pixel 7 Pro. Would not be as disorienting if there were a way to disable or adjust animation/motion settings.

**Comment by exstntlmsnthrp** [comment]
Interesting. I'm using the Pixel 7 Pro and I do not seem to notice these problems. I've used it on both stock Android and I'm currently using LineageOS with no problem.

**Comment by russianbot91** [comment]
I think a good solution would be the ability to change the animation speed within the launcher settings.

I use a Pixel 6a and have the animation scale set to 0.5x in Developer Settings, but I think the animations in Kvaesitso could be sped up.

**Comment by rohitghali** [comment]
Yep. Can confirm. Suffering the same here. It's like scrolling at low refresh rate. OnePlus 7t pro here.

**Comment by kinesight** [comment]
Same issue here on Samsung Galaxy A52s - OneUI 4.1 (Android 12)

**Comment by MM2-0** [comment]
Can you check if that build improves anything for you?

[app-default-release.zip](https://github.com/MM2-0/Kvaesitso/files/11301532/app-default-release.zip)

**Comment by Terkyz** [comment]
I have been testing it and the difference is [noticeably better](https://imgur.com/a/c2m159j). At first for some reason (I do not know if it is related to the launcher) the keyboard seems to take a while to appear and the apps at first have that effect of "heaviness" but really after a moment and goes very smooth and fast.

**Comment by MM2-0** [comment]
> the apps at first have that effect of "heaviness" but really after a moment and goes very smooth and fast.

I have improved the caching so resources are reused more efficiently, but the first load remains unchanges so that's to be expected.

**Comment by Terkyz** [comment]
The themed icons are disabled by default in my system, so it can be discarded as an option.

I tried different icon shapes and I definitely noticed a change, I have the feeling that the square is the fastest, while the circle is pretty fast too.

Once you have the icons in circular shape and the labels disabled, it is VERY FLUID.

Lowering the number of columns with the previous settings applied does not represent any noticeable change, I went back to the default settings and it does feel a change, mainly with 3 columns. 

The themed icons are disabled by default in my system, so it can be discarded as an option.

I tried different icon shapes and I definitely noticed a change, I have the feeling that the square is the fastest, while the circle is pretty fast too.

Once you have the icons in circular shape and the labels disabled, it is VERY FLUID.

Lowering the number of columns with the previous settings applied does not represent any noticeable change, I went back to the default settings and it does feel a change, mainly with 3 columns. 

[Here](https://imgur.com/a/4zWy8x4)  I edited a short video where you can better appreciate the performance difference, I could not add when I change the number of columns due to lack of space unfortunately.

**Comment by yant0** [comment]
I face the same issue yesterday, i decide to open up the apps's setting and i forgot that i had battery optimization on. After turning it off, its running smooth although still choppy (~18fps). I have an android 11 sd600 3gb ram.

**Comment by Akselmo** [comment]
I have same issue on Fairphone 3+. If I could disable the physics based animations and translucency in cards, I think that would help.

**Comment by dongltron** [comment]
Hey there, I can also confirm the laggy performance on Pixel 7 Pro. It's not awful but definitely noticeable at 120Hz. Disabling labels seems to help (that's how I prefer it anyway), but still definitely a noticeable drop in performance. If I can help in any way let me know, totally willing to test a build or troubleshoot :)

**Comment by dongltron** [comment]
> Hey there, I can also confirm the laggy performance on Pixel 7 Pro. It's not awful but definitely noticeable at 120Hz. Disabling labels seems to help (that's how I prefer it anyway), but still definitely a noticeable drop in performance. If I can help in any way let me know, totally willing to test a build or troubleshoot :)

https://imgur.com/a/SmqUnyg

Here's a link to my experience. Definitely not as bad as the S10 example above but still notable.

**Comment by dongltron** [comment]
Update: So, to me, the poor performance appears to be something to do with the keyboard? Upon updating to the latest version, 1.25.0, the performance has straightened itself out, but only when the keyboard is closed. Upon opening it (therefore when the keyboard opens) the performance is laggy as usual. As soon as it is closed, to my eyes at least, it's 120Hz.

Video provided below. Hopefully this helps to narrow down the issue?

https://imgur.com/a/6T4tjlK

**Comment by dongltron** [comment]
Disabling 'Open keyboard' fixes the lag entirely. I'm almost certain it's something to do with the keyboard itself.

**Comment by Akselmo** [comment]
> Disabling 'Open keyboard' fixes the lag entirely. I'm almost certain it's something to do with the keyboard itself.

Unfortunately not on Fairphone 3+. But this phone has way less resources anyhow.

**Comment by dongltron** [comment]
Ah okay yeah I didn't mention it here because I did above but just in case people miss it this is on a Pixel 7 Pro, so yeah definitely punching above the Fairphone's weight. Hopefully not too hard to fix once they pinpoint the issue. Seems they have taken the necessary steps for the rest of the scrolling though, just the keyboard left to optimise (I think?)

**Comment by MM2-0** [comment]
I think the keyboard issue is somehow related to this: #370

**Comment by projjalm** [comment]
I think the cause(s) of this FPS drop are not one, but many, because I've been testing the app recently and noticed it happens under very different situations.  

It first happens when you start the app after a force stop (or after other actions I didn't test with it yet) and pull up/down to the drawer/widgets WITHOUT waiting for some seconds (to let it cache stuff?). It generally happens whenever app icons are getting loaded, that is, on app start in favorites row or drawer.  

Then, most importantly, **it happens when the keyboard gets enabled while in the drawer** (synonymous to the search box getting clicked).  

And also about the linked issue, @MM2-0 while the behavior (glitch) of the keyboard maybe linked to the above cause, the FPS drop happens regardless of the glitch (that is, keyboard opening and closing in a short period of time).

**Comment by MM2-0** [comment]
> And also about the linked issue, @MM2-0 while the behavior (glitch) of the keyboard maybe linked to the above cause, the FPS drop happens regardless of the glitch (that is, keyboard opening and closing in a short period of time).

It's just a theory, but maybe something triggers the keyboard to show and hide multiple times in a row which

1. Forces the layout to remeasure everytime the keyboard state changes (hence the lag)
2. Can produce the visual glitch that is described in the other issue (however I believe that the glitch is not always visible as it might depend on other factors as well, e.g. how long the keyboard takes to open)

**Comment by mjain234** [comment]
First, I want to give thanks to the developer for this awesome launcher he built and hope so he still continues to do till infinity. I also faced a lag while doing a search in app drawer. I have read all the comments here and followed the same by disabling labels, choose icon shape to circle too and change the column to 3 but the lag is still noticible as you can see [here](https://i.imgur.com/hrb0LBn.mp4).

1)Whenever swiping and keyboard shows, there's a small lag noticeable, also 
2)when you scroll up in app drawer and also
3) sometimes when you exit app and go to home screen.

**Comment by mjain234** [comment]
> First, I want to give thanks to the developer for this awesome launcher he built and hope so he still continues to do till infinity. I also faced a lag while doing a search in app drawer. I have read all the comments here and followed the same by disabling labels, choose icon shape to circle too and change the column to 3 but the lag is still noticible as you can see [here](https://i.imgur.com/hrb0LBn.mp4).
> 
> 1)Whenever swiping and keyboard shows, there's a small lag noticeable, also 
> 2)when you scroll up in app drawer and also
> 3) sometimes when you exit app and go to home screen.
> 
> 

Also, while in home screen, there are some scrollable widgets where I am not able to scroll properly, whenever I do instead of scrolling in the widgets itself, it gets scrolled in main screen itself. See [here](https://i.imgur.com/PmYBUPK.mp4)

**Comment by dikey0ficial** [comment]
I've faced this bug today after firmware update on my Honor X9a. Disabling keyboard launching helped, but when I clicking at search bar from main screen or app drawer this lag is notable.

Huge gratitude for developing Kvaesito and maintaining it!;)

**Comment by rahaaatul** [comment]
> Disabling keyboard launching helped, but when I clicking at search bar from main screen or app drawer this lag is notable.

I can confirm this. It does the same thing in my device, here's a screen record @MM2-0 :

https://github.com/MM2-0/Kvaesitso/assets/86149861/67bd6829-4c59-49d0-b3a2-de6adeb98afa

**Comment by gromoslaw-kroczka** [comment]
Could be related to #577 ?

**Comment by dongltron** [comment]
Happy to report that 1.28.0 appears to have completed fixed these issues.

Pixel 8 Pro on Android 14

**Comment by TheDeadCell** [comment]
After 1.28.0, still experiencing this issue on Oneplus Nord N10 android 11.

**Comment by Sir-Photch** [comment]
Unfortunately, since GrapheneOS moved to Android 14, Kvaesitso just gets super laggy after each update. While reinstalling the app fixes that problem, this is not really a solution.

**Comment by Rolv-Apneseth** [comment]
I'm also on GrapheneOS and it's the same deal for me, app only gets super laggy after an update and re-installing fixes it. I was wondering if it was something to do with the app optimisation it does after each update but maybe not if a bunch of other phones are also having the same issue?

**Comment by exstntlmsnthrp** [comment]
This must have to do with grapheneos and the post update optimizations that others reference. To my knowledge, every variation of Android does these optimizations to some extent or other after each update. I also have a Pixel 7 pro, but I have LineageOS installed and do not have this problem. 

To be fair, I have noticed issues with other apps after system updates, just not kvaesitso.

**Comment by Sir-Photch** [comment]
Instead of reinstalling on GrapheneOS, you may also try to switch back to the default launcher, disable Kvaesitso and clear it's cache, restart your phone for GOS to re-optimize, and then re-enable it as a launcher. This will keep your configuration (esp. the non-exportable).

Seems to me that apps that have access to gestures / are set as a launcher do not get optimized.

**Comment by Sir-Photch** [comment]
IIRC this will remove your configuration of the launcher.

**Comment by Sir-Photch** [comment]
Actually, this may be fixed for GOS users since the latest update. (https://github.com/GrapheneOS/os-issue-tracker/issues/2526) Can anyone else on GrapheneOS confirm this? It seems fine for me now.

**Comment by exstntlmsnthrp** [comment]
@Sir-Photch I agree with you. After I got the update yesterday, I've noticed that I never had to force stop, clear cache and reboot. I was wondering if something changed it it was just me.

**Comment by Rolv-Apneseth** [comment]
Same for me when updating yesterday, but I didn't notice an app optimising step either

**Comment by jacekpoz** [comment]
it has been better for me recently (latest GOS alpha) but there's still noticable lag, I don't think it's fixed yet

force stopping, clearing cache and rebooting also doesn't help

**Comment by Rolv-Apneseth** [comment]
Just did an update and the issue remains for me. This one did have the app optimising step.

**Comment by Pentaphon** [comment]
The app drawer is easily the slowest part of Kvaesitso even on the current version posted today.

**Comment by Rolv-Apneseth** [comment]
The problem described by this issue is solved I think, at least for me it's fine now. I'm sure the performance could be better, but imo it makes sense that the drawer is slower than anything else, it has to display the most.

**Comment by rahaaatul** [comment]
> imo it makes sense that the drawer is slower than anything else, it has to display the most.

Use other app launchers and tell me if there's a slowdown in app drawer @Rolv-Apneseth. It's simply a coding issue somewhere.

**Comment by Rolv-Apneseth** [comment]
Fair. I guess it is good enough for me but is an issue for others

**Comment by Pentaphon** [comment]
@MM2-0 Perhaps the performance issues are due to the app not being 100% Kotlin?

**Comment by voytxt** [comment]
Over the last couple updates, this issue has completely disappeared for me! (I'm on Samsung A52)

**Comment by ImranR98** [comment]
I'm seeing this on a fresh install of the latest release. Pixel 9 Pro, GrapheneOS.
